### DISCUSSION TITLE
WiX v4: CustomAction Type 21/22 syntax

### DISCUSSION THREAD

**Discussion by Garrkrom** [wixtoolset/issues#7507] [Questions]
Hi,

I want to move my projects from WiX v3 to WiX v4. I have some custom VB script actions (Type 37 or 38). Now I'm not able to find the correct XML syntax for this. Maybe this feature is still missing in WiX v4?

I thought the following XML should be correct. But it does not work. Visual Studio states the text must be empty.

`<CustomAction Id="MyActionId" Script="vbscript"><![CDATA[
MyCustomScript
]]></CustomAction>`

In the corresponding xsd file the documentation states, that the script attribute needs the script in the element's text.
Has anyone had success with this?

**Comment by barnson** [comment]
Type 21/22 custom actions use an installed file. Type 37/38 CAs use embedded script. https://wixtoolset.org/docs/schema/wxs/customaction/ says:

> Script (enumeration) : Creates a type 37 or 38 custom action. Specify a path to the script to be embedded in the package in the ScriptSourceFile attribute.

**Reply by mvollmer1** [reply]
Using this syntax, my VS2022 build using v4.0.1, tells me "**The 'ScriptSourceFile' attribute is not allowed**."   It worked find in v3.  

The element for custom action has ID, script set to vbscript and the sourcescriptfile attributes.

<CustomAction ID="myid" Script="vbscript" ScriptSourceFile="location of my vbs file." />

**Reply by mvollmer1** [reply]
Do we know if this is a bug or if the manual needs to be corrected?

**Reply by mvollmer1** [reply]
<CustomAction Id="MyActionId" Script="vbscript" ScriptSourceFile="myscriptfile.vbs" />

"The 'ScriptSourceFile' attribute is not allowed."

**Reply by mvollmer1** [reply]
Script (enumeration) : Creates a type 37 or 38 custom action. Specify a path to the script to be embedded in the package in the ScriptSourceFile attribute. This attribute's value must be one of the following:

    jscript
    vbscript

ScriptSourceFile (String) : Path to the external file containing the script code. Can be used only with the Script attribute.

**Reply by robmen** [reply]
You do have access to the source code: https://github.com/wixtoolset/wix/blob/f18d7e628361cd3f2ff14dd3326cbeea68df004f/src/wix/WixToolset.Core/Compiler.cs#L3388

**Reply by mvollmer1** [reply]
As I work with the solution and build, I'm not yet seeing this error in the error list.   But it shows as a red squiggly line in the UI.   I'm wondering now, if the problem isn't with Wix, but with Heatwave.

**Reply by robmen** [reply]
Ahh, well, that is an important distinction. HeatWave is due for an update in the XSDs. It is possible this one is out of date. We (at FireGiant) are looking at doing a HeatWave refresh a bit after the time WiX v4.0.2 comes out.
